### PR TITLE
Task: cache autocomplete results

### LINKED ISSUE #724: Autocomplete should cache URL patterns
**Describe the bug**
Autocomplete doesn't cache the results for a URL pattern. I see it making repeat APIs calls for the same URL pattern. 

**To Reproduce**
1. Open Fiddler, capture
2. Type `https://graph.microsoft.com/v1.0/users/123/`, delete `123/`, the add `123/`.

**Expected behavior**
No new API call is made as we already downloaded the autocomplete suggestions for that URL pattern.
 
[AB#6318](https://microsoftgraph.visualstudio.com/0985d294-5762-4bc2-a565-161ef349ca3e/_workitems/edit/6318)

### PR DESCRIPTION
**Pull request by thewahome**
## Overview

Uses indexed db to store the results of the autocomplete query so that subsequent calls do not call the API. This effectively speeds up autocomplete triggers based on the same url

Fixes #724

### Demo

image

### Notes

Request URLs with keys are replaced with generic IDs so that they are easily recreated for storage and searching in the cache.

### Tasks

- [x] Create store for autocomplete options
- [x] Read and write options to the store
- [x] Add a time to live for the options in the store

## Testing Instructions

* Run the app. Trigger an autocomplete search e.g `me/messages?`
* Using the developer console, notice that the request URL is placed as a key in the indexedDb storage
* perform another autocomplete search e.g `me/todo?`
* Using the developer console, notice that there was a network request that was made
* Redo `me/messages?` and notice that it is no longer calling the API

### CODE REVIEW

File: src/app/views/query-runner/query-input/auto-complete/auto-complete.cache.ts
Comment (jobala) [inline src/app/views/query-runner/query-input/auto-complete/auto-complete.cache.ts:13]: Specify return type.

File: src/app/views/query-runner/query-input/auto-complete/auto-complete.cache.ts
Comment (jobala) [inline src/app/views/query-runner/query-input/auto-complete/auto-complete.cache.ts:32]: I think you can leave out the type here to be consistent or specify the type for all variable assignments.

**Review by jobala — CHANGES_REQUESTED** [review]
It is possible to improve storage performance by storing the hash of a path instead of the path itself because in the worst case scenario the path can be really long.

**Comment by jobala** [comment]
#### Design Proposal

@thewahome what do you think of abstracting the logic for fetching suggestions into a SuggestionProvider module that has the simple interface `getSuggestion(...)` then under the hood it will figure out whether to get the suggestion from cache or network.

Ideally, someone else with a similar use case should be able to use this module without understanding how it works.

Let me know what you think.

**Comment by thewahome** [comment]
> #### Design Proposal
> @thewahome what do you think of abstracting the logic for fetching suggestions into a SuggestionProvider module that has the simple interface `getSuggestion(...)` then under the hood it will figure out whether to get the suggestion from cache or network.
> 
> Ideally, someone else with a similar use case should be able to use this module without understanding how it works.
> 
> Let me know what you think.

@jobala this sounds good and worth looking into... We can track this as technical debt?

**Comment by jobala** [comment]
> @jobala this sounds good and worth looking into... We can track this as technical debt?

@thewahome if it was up to me I would say we do this before merging. I will leave it up to @ddyett and @MIchaelMainer to make the call.

File: src/modules/suggestions/suggestions.ts
Comment (jobala) [inline src/modules/suggestions/suggestions.ts:20]: This method should be private.

**Review by jobala — APPROVED** [review]

**Pull request merged**
